### PR TITLE
Device consolidation

### PR DESCRIPTION
**Pull request by rm-hull**

### CODE REVIEW

**Comment by thijstriemstra** [comment]
@rm-hull what about having a package called `luma.oled.device` with base stuff in `__init__` and then a `color`/`grayscale` module.

**Comment by rm-hull** [comment]
Yep, that could work, I think ... so to clarify:
* `luma/oled/device/__init__.py` <-- all existing classes in here
* `luma/oled/device/color.py` <-- color base code in here
* `luma/oled/device/greyscale.py` <-- greyscale base code in here

?

**Comment by thijstriemstra** [comment]
Yep.

**Comment by rm-hull** [comment]
Am surprised this caused doc tox to fail. Can’t quite fathom the reason why though

**Comment by thijstriemstra** [comment]
Yea that's weird, I'll take a look later today.

**Comment by thijstriemstra** [comment]
@rm-hull fixed the (setuptools related) issue and updated some documentation. Also added version marker for '3.0.0', seems like a good nxt version..

File: luma/oled/device/greyscale.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                    buf[i // 2] |= grey
+
+            i += 1
+
+    def display(self, image):
+        """
+        Takes a 1-bit monochrome or 24-bit RGB image and renders it
+        to the SSD1322 OLED display. RGB pixels are converted to 4-bit
+        greyscale values using a simplified Luma calculation, based on
+        *Y'=0.299R'+0.587G'+0.114B'*.
```
Comment (thijstriemstra) [inline luma/oled/device/greyscale.py]: this docstring needs an update..

Comment (rm-hull) [reply luma/oled/device/greyscale.py]: fixed

**Comment by thijstriemstra** [comment]
Can you also bring back coverage to a 100% with testing these `NotImplementedErrors`? And ssd1351 also has a new untested large branch.

**Comment by thijstriemstra** [comment]
And the tests can probably also be simplified?

**Comment by rm-hull** [comment]
Made those methods abstract, so not really possible to instantiate a subclass without a concrete implementation now, so just marked them as `#pragma: no cover`.

Consciously didn't want to change the tests in the same PR in order to 'prove' that there was no regression in behaviour.

**Comment by thijstriemstra** [comment]
@rm-hull perhaps some general docstring to describe the intention and params of these methods?

**Comment by rm-hull** [comment]
Not sure about commit 58d92b3 - did it so that the class hierarchy looks a bit tighter

image

Thinking the greyscale and color modules could be pushed up a level so that they're not under device. 

WDYT?

**Comment by thijstriemstra** [comment]
I'm more a fan of `luma.oled.device.color.color_device` tbh (and not use same classname twice, or even three times if you count luma.core's device as well). So you import `color_device` in `luma.oled.device` (from `luma.oled.device.color`).

**Comment by thijstriemstra** [comment]
I first thought you were going to move the relevant color devices into `color_device` module and then import them into `luma.oled.device` so we can keep that `__all__` intact @rm-hull

**Comment by thijstriemstra** [comment]
I also suggest to remove the long lists of device references and replace it with something like "for color and greyscale devices' (or monochrome?). Because this list will keep on growing ;)

File: luma/oled/device/color.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        self._init_sequence()
+        self.contrast(0xFF)
+        self.clear()
+        self.show()
+
+    @abstractmethod
+    def _supported_dimensions(self):
+        """
+        Enumerates the screen resolutions that the specific device supports, as
+        a list of tuples; e.g.: [(96, 64), (96, 32), (96, 16)]
```
Comment (thijstriemstra) [inline luma/oled/device/color.py]: should be wrapped, e.g.
```
``[(96, 64), (96, 32), (96, 16)]``
```

**Comment by rm-hull** [comment]
Hey @thijstriemstra, can we push PR this through? Can quite fathom what was blocking it ... maybe a fresh review? 😬

**Comment by thijstriemstra** [comment]
Sounds good, will take another look asap.

**Comment by thijstriemstra** [comment]
Will do!

File: luma/oled/device/color.py (hunk 2 above)
Comment (thijstriemstra) [reply luma/oled/device/color.py]: ping @rm-hull

Comment (rm-hull) [reply luma/oled/device/color.py]: This was done ages ago - see https://github.com/rm-hull/luma.oled/pull/203/files#diff-12a7127a588670816142a0e5d0a15c1cR40 - the snippet above is showing as 'outdated'

**Pull request merged**